src/room_member.cpp

```diff
@@ -0,0 +1,200 @@
+            case ID_ROOM_NAME_COLLISION:
+                state = State::NameCollision;
+                peer->Shutdown(300);
+                break;
+            case ID_ROOM_MAC_COLLISION:
+                state = State::MacCollision;
+                peer->Shutdown(300);
+                break;
+            case ID_ROOM_JOIN_SUCCESS:
+                // The join request was successful, we are now in the room.
+                // If we joined successfully, there must be at least one client in the room: us.
+                ASSERT_MSG(GetMemberInformation().size() > 0, "We have not yet received member information.");
+                state = State::Joined;
+                break;
+            case ID_DISCONNECTION_NOTIFICATION:
+                // Connection lost normally
+                state = State::Idle;
+                peer->Shutdown(300);
+                break;
+            case ID_INCOMPATIBLE_PROTOCOL_VERSION:
+                state = State::WrongVersion;
+                peer->Shutdown(300);
+                break;
+            case ID_CONNECTION_ATTEMPT_FAILED:
+                state = State::Error;
+                peer->Shutdown(300);
+                break;
+            case ID_NO_FREE_INCOMING_CONNECTIONS:
+                state = State::RoomFull;
+                peer->Shutdown(300);
+                break;
+            case ID_CONNECTION_LOST:
+                // Couldn't deliver a reliable packet, the other system was abnormally terminated
+                state = State::LostConnection;
+                peer->Shutdown(300);
+                break;
+            case ID_CONNECTION_REQUEST_ACCEPTED:
+                // Update the server address with the address of the sender of this packet.
+                server_address = packet->systemAddress;
+                // TODO(Subv): Send a room join request to the server before marking us as joined.
+                break;
+            default:
+                break;
+            }
+
+            peer->DeallocatePacket(packet);
+        }
+    }
+};
+
+void RoomMember::Join(const std::string& nickname, const std::string& server, uint16_t server_port, uint16_t client_port) {
+    // TODO(Subv): Use client_port.
+    RakNet::SocketDescriptor socket;
+    peer->Startup(1, &socket, 1);
+
+    RakNet::ConnectionAttemptResult result = peer->Connect(server.c_str(), server_port, nullptr, 0);
+    if (result != RakNet::CONNECTION_ATTEMPT_STARTED) {
+        state = State::Error;
+        return;
+    }
+
+    // TODO(Subv): Keep track of the desired nickname
+    state = State::Joining;
+
+    // Start a network thread to Receive packets in a loop.
+    receive_thread = std::make_unique<std::thread>(&RoomMember::ReceiveLoop, this);
+}
+
+bool RoomMember::IsConnected() const {
+    return state == State::Joining || state == State::Joined;
+}
+
+void RoomMember::Leave() {
+    ASSERT_MSG(receive_thread != nullptr, "Must be in a room to leave it.");
+
+    peer->Shutdown(300);
```

Comment:
A possible fix would be to have a lock here and inside the while (IsConnected()) loop